src/main/java/org/opensearch/geospatial/ip2geo/cache/GeoDataCache.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.geospatial.ip2geo.cache;
+
+import java.net.InetAddress;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Function;
+
+import lombok.AllArgsConstructor;
+import lombok.EqualsAndHashCode;
+
+import org.opensearch.common.cache.Cache;
+import org.opensearch.common.cache.CacheBuilder;
+
+/**
+ * Cache to hold geo data
+ */
+public class GeoDataCache {
+    private Cache<CacheKey, Map<String, Object>> cache;
+
+    public GeoDataCache(final long maxSize) {
+        if (maxSize < 0) {
+            throw new IllegalArgumentException("ip2geo max cache size must be 0 or greater");
+        }
+        this.cache = CacheBuilder.<CacheKey, Map<String, Object>>builder().setMaximumWeight(maxSize).build();
+    }
+
+    public Map<String, Object> putIfAbsent(
+        final String datasourceName,
+        final InetAddress ip,
+        final Function<InetAddress, Map<String, Object>> retrieveFunction
+    ) throws ExecutionException {
+        CacheKey cacheKey = new CacheKey(datasourceName, ip);
+        return cache.computeIfAbsent(cacheKey, key -> retrieveFunction.apply(key.ip));
+    }
+
+    public Map<String, Object> get(final String datasourceName, final InetAddress ip) {
+        return cache.get(new CacheKey(datasourceName, ip));
+    }
+
+    /**
+     * Create a new cache with give size and replace existing cache
+     *
+     * @param maxSize
+     */
+    public void updateMaxSize(final long maxSize) {
```

Comment:
This will clear existing cache. is this expected?

Reply:
Yes. On a second thought, let me try to copy the existing value from old to new cache.